### PR TITLE
CE3 Testing upgrade

### PR DESCRIPTION
**Pull request by domaspoliakas**
Hello,

Upgraded testing. Mostly involved adding ways for IO to be run unsafely.

One thing that I found peculiar was the resource usage in `Http4sSpec.withResource`. I couldn't quite understand why it wasn't just `.use` there before, but for now I just replaced it with that since `Resource` has been reworked and the old code doesn't work for it anymore anyway.

### CODE REVIEW

File: testing/src/test/scala/org/http4s/Http4sSpec.scala [hunk 1]
```diff
@@ -99,18 +98,7 @@ trait Http4sSpec
...
-          .map { case (a, release) =>
-            fs(a).append(step(release(ExitCase.Completed).unsafeRunSync()))
-          }
-          .unsafeRunSync()
-      case Resource.Bind(r, f) =>
-        withResource(r)(a => withResource(f(a))(fs))
-      case Resource.Suspend(r) =>
-        withResource(r.unsafeRunSync() /* ouch */ )(fs)
-    }
+    r.use(a => IO(fs(a))).unsafeRunSync()
```
Comment (rossabaker) [inline testing/src/test/scala/org/http4s/Http4sSpec.scala:101]: I think this is going to provide a released `a` to the fragments when they run.  Note how the old implementation was appending a step to the returned fragments.

If we can't use this allocate backdoor anymore, we're going to have a struggle in a couple modules.

**Review by rossabaker — CHANGES_REQUESTED** [review]
Great start, but we're going to need to find another way for resourceful testing.

File: testing/src/test/scala/org/http4s/Http4sSpec.scala (hunk 1 above)
Comment (hamnis) [reply testing/src/test/scala/org/http4s/Http4sSpec.scala:101]: is the allocated method gone in cs3 completely?

Comment (hamnis) [reply testing/src/test/scala/org/http4s/Http4sSpec.scala:101]: if not, could we use that to append the step?

Comment (domaspoliakas) [reply testing/src/test/scala/org/http4s/Http4sSpec.scala:101]: `allocated` is still around. 

I'll go back to it and figure out a way 👍

File: testing/src/test/scala/org/http4s/Http4sSpec.scala [hunk 2]
```diff
@@ -98,7 +98,8 @@ trait Http4sSpec
   }
 
   def withResource[A](r: Resource[IO, A])(fs: A => Fragments): Fragments =
-    r.use(a => IO(fs(a))).unsafeRunSync()
+    r.allocated.map { case (r, release) => fs(r).append(step(release.unsafeRunSync())) }.unsafeRunSync()
+    // r.use(a => IO(fs(a))).unsafeRunSync()
```
Comment (domaspoliakas) [inline testing/src/test/scala/org/http4s/Http4sSpec.scala:102]: One day I'll remember to clean up before pushing I swear

**Comment by rossabaker** [comment]
Needs a scalafmt, and I think it's ready.

**Comment by domaspoliakas** [comment]
Did that. I also noticed CI was complaining about cats effect laws not necessary for laws and testing, so I removed that too.

**Comment by rossabaker** [comment]
Oh, I discovered cleaned that up in #3832. Sorry, should have tagged you.

**Comment by rossabaker** [comment]
Keep your version of the merge conflict, and I think we're all set.

**Comment by domaspoliakas** [comment]
All done now

File: build.sbt [hunk 3]
```diff
@@ -94,6 +94,10 @@ lazy val laws = libraryProject("laws")
     description := "Instances and laws for testing http4s code",
     libraryDependencies ++= Seq(
       caseInsensitiveTesting,
+<<<<<<< HEAD
+=======
+      catsLaws,
+>>>>>>> upstream/cats-effect-3
```
Comment (rossabaker) [inline build.sbt:100]: Keep `catsLaws`, lose the conflict markers.

**Comment by domaspoliakas** [comment]
Apologies for the sloppy mess of a PR. I am very tired 😄

**Comment by rossabaker** [comment]
We have 9.843 commits on main.  It's not the first mess!  You're doing great work, and this is the one that will open up a handful more modules.

**Comment by rossabaker** [comment]
Build failure was a CI infrastructure flake. Restarting.

**Review by rossabaker — APPROVED** [review]
:+1: on green

**Pull request merged**